### ISSUE TITLE
[Ivy AOT] AOT recompilation is working slow on big projects

### ISSUE THREAD

**Issue by nickbullock** [angular/angular#35906]
# 🐞 bug report

### Affected Package

 The issue is caused by package @angular/compiler-cli

### Is this a regression?

 No

### Description

We have a big Angular Ivy project, and the problem is that we want to use Ivy+AOT everywhere: in production and development, but we can't use it for development now because of slow recompilation AOT process. 
Project is `@nrwl/nx` based app with > 10 libraries and two applications.
Also can be important: `ng-cli` based with custom webpack extensions, hybrid app with angularjs 1.5.8.

Some stats (Ivy+AOT):

Full application:
- first compile: 96225ms - 180000ms
- recompile: 11949ms (+-) - 20000ms

Demo application (small part of full application):
- first compile: 35987ms 
- recompile: 3040ms

AngularCompilerPlugin stats (in full app):
- 3915 files reported by NgTscProgram
- 2152 classes reported by TraitCompiler

AngularCompilerPlugin stats (in demo app):
- 1600 files reported by NgTscProgram

JIT is working much faster and recompilations tooks always 2-4sec.

I turned on benchmark in node_modules/@ngtools/webpack/src/benchmark.js and started logging why AngularCompilerPlugin works so long on recompilations. Then I added my own logs prefixed with `>>>`. I added these log files as assignments to this issue. Log files contains our full app compilation and recompilation process - it separated with `[RECOMPILE STARTS HERE]` tag.

In the first log file you can see that chain
```AngularCompilerPlugin._make => AngularCompilerPlugin._update => AngularCompilerPlugin._createOrUpdateProgram.ng.loadNgStructureAsync ```
take 10311.904ms it is about 10/17 of all recompilation time.
Then dived deep into it and understood that `ComponentDecoratorHandler.resolve` is the slow method
using console.time. Here is stacktrace:

```
ComponentDecoratorHandler.resolve (component.js:382)
TraitCompiler.resolve (compilation.js:395)
NgCompiler.resolveCompilation (compiler.js:356)
NgCompiler.analyzeAsync (compiler.js:181)
NgtscProgram.loadNgStructureAsync (program.js:142)
_createOrUpdateProgram (angular_compiler_plugin.js:261)
_update (angular_compiler_plugin.js:842)
_make (angular_compiler_plugin.js:749)
```
In the second log you can see 3 internal timelines of `ComponentDecoratorHandler.resolve()` and here is how I did it:
usedDirectives
![image](https://user-images.githubusercontent.com/15385798/76080339-bb0bfb80-5fb7-11ea-938c-0e16913f6535.png)
usedPipes
![image](https://user-images.githubusercontent.com/15385798/76080405-e0006e80-5fb7-11ea-9731-a6eabbb3bd01.png)
wrapDirectivesAndPipesInClosure
![image](https://user-images.githubusercontent.com/15385798/76080450-fa3a4c80-5fb7-11ea-8632-ab9f5e09e5ba.png)

Sometimes `ComponentDecoratorHandler.resolve.usedDirectives` and  `ComponentDecoratorHandler.resolve.usedPipes` can execute very long time (check it in log file):
```
>>>> ComponentDecoratorHandler.resolve.usedDirectives: 23.926ms
>>>> ComponentDecoratorHandler.resolve.usedDirectives: 65.800ms
>>>> ComponentDecoratorHandler.resolve.usedPipes: 11.891ms
>>>> ComponentDecoratorHandler.resolve.usedPipes: 33.837ms
```
These function called about ~2000 times in our app recompilation process so we can understand that if one usedDirectives and usedPipes processing tooks for example 5ms then whole process will took about 5x2x2000 = 20000ms = 10s.

In our application usedDirectives variable in `ComponentDecoratorHandler.resolve` method was equal to the value of 43 (maximum) and usedPipes to the value 7 (maximum).

tsc --listFiles says there are 3913 files in typescript compilation. Added files list as assignment.

Main part of tsconfig:
```
{
  "compilerOptions": {
    "types": [
      "@types/angular",
      "@types/chart.js",
      "@types/core-js",
      "@types/jasmine",
      "@types/jquery",
      "@types/lodash",
      "@types/lru-cache",
      "@types/node",
      "@types/ui-select"
    ],
    "rootDir": ".",
    "baseUrl": ".",
    "moduleResolution": "node",
    "lib": ["es2018", "dom"],
    "module": "esnext",
    "sourceMap": true,
    "allowJs": false,
    "experimentalDecorators": true,
    "esModuleInterop": true,
    "target": "es5",
    "resolveJsonModule": true,
    "allowSyntheticDefaultImports": true,
    "declaration": false
  },
  "angularCompilerOptions": {
    "skipMetadataEmit": true,
    "enableIvy": true,
    "fullTemplateTypeCheck": true
  }
}

```

I think that any big project can get into this issue with grow of files count which are included to the typescript and angular compilation processes.

What I can't understand here: why Angular needs to traverse and handle all typescript compilation files when changed files (usually it one file) are known already )? It is looks like Ivy AOT can't work in incremental mode? How we can boost AOT recompilation without reducing files count?

Please tell me if you will need something else to understand the problem.

## 🔬 Minimal Reproduction

 It is very hard to reproduce it on stackblitz or github

## 🌍  Your Environment

**Angular Version:**
<pre><code>

Angular CLI: 9.0.4
Node: 12.13.1
OS: darwin x64

Angular: 9.0.4
... cli, common, compiler, compiler-cli, core, forms
... platform-browser, platform-browser-dynamic, upgrade
Ivy Workspace: Yes

Package                           Version
-----------------------------------------------------------
@angular-devkit/architect         0.900.1
@angular-devkit/build-angular     0.900.4
@angular-devkit/build-optimizer   0.900.4
@angular-devkit/build-webpack     0.900.4
@angular-devkit/core              9.0.4
@angular-devkit/schematics        9.0.1
@angular/cdk                      9.1.0
@ngtools/webpack                  9.0.4
@schematics/angular               9.0.4
@schematics/update                0.900.4

rxjs                              6.5.3
typescript                        3.7.5
webpack                           4.41.2

</code></pre>

**Anything else relevant?**

Log1
[1.log-first-compile.txt](https://github.com/angular/angular/files/4298250/1.log-first-compile.txt)
Log2
[2.log-comp-dec-resolve-used-dirs-and-pipes.txt](https://github.com/angular/angular/files/4298249/2.log-comp-dec-resolve-used-dirs-and-pipes.txt)
tsc list files
[tsc-list-files.txt](https://github.com/angular/angular/files/4298251/tsc-list-files.txt)

**Comment by JoostK** [comment]
Thank you very much for the detailed report! I have seen similar performance bottlenecks in earlier reports, e.g. #34867. The cyclic import computation is particularly slow as it is module resolution heavy, and therefore I/O intensive.

**Comment by sod** [comment]
In our app (190 components) touching any typescript file thats part of the shared module (imported by every other module) is very slow (12 seconds). Anything else (any template, any scss file, any part of a lazy module) is below two seconds.

In our case it's basically this: https://github.com/angular/angular/blob/4253662231f8e3f6de137ab511ff7fce4871a168/packages/compiler-cli/src/ngtsc/core/src/compiler.ts#L512-L514
that invalidates the entire app, as the shared module is imported everywhere.

I wish angular would allow components to be root (just like services now) and components directly imported where they used (just like vue or react). Then incremental compilation doesn't have to aggressively invalidate entire blocks of code and code splitting is way easier.

**Comment by JoostK** [comment]
@sod incremental compilation is not _required_ to aggressively invalidate lots of files, even with the current design, _however_ determining when to invalidate which files is a hard problem that is yet to be explored. There's certainly optimizations possible where the "API surface" of an NgModule is tracked and compared to the previous compilation, such that changes that do not affect the API surface don't have to invalidate the related transitive dependencies.

You can find some details in [the README of the incremental engine](https://github.com/angular/angular/blob/master/packages/compiler-cli/src/ngtsc/incremental/README.md#what-optimizations-are-possible-in-the-future).

**Comment by bkarv** [comment]
Just upgraded from Angular 8 to Angular 9 for a large application 100+ components. Although it's not as slow as your compilation I have noticed an increase in AOT time, estimated 20%.

**Comment by alxhub** [comment]
I suspect this is fixed (or at least significantly improved) in Angular 10 via ecffc3557.

**Comment by JoostK** [comment]
Hi @nickbullock,

could you please try upgrading to 10.1.2 and report on how it performs for you? It contains a number of performance improvements that should be noticeable in large projects like yours. I am a bit surprised by the parts that appear to be slow for you, as I have tested with a couple large projects and that particular code path is not slow for me (it's typically the computation of `cycleDetected` just before your logging where things are slow). I'd be interested to profile using your app, if you are able to share the repo privately.

**Comment by sod** [comment]
Seeing ~10% faster compile times for us in dev mode - running in aot mode

 x                              | startup | 1. change | 2. change | 3. change | 4. change 
------------------ | ------- | ---------- | ---------- | ---------- | ----------
10.1.0 component typescript   | 59.0s   | 12.4s     | 10.6s     | 8.8s      | 8.8s      
10.1.2 component typescript   | 53.6s   | 11.0s     |  9.3s     | 9.0s      | 8.4s      

 component template                              | startup | 1. change | 2. change | 3. change | 4. change 
------------------ | ------- | ---------- | ---------- | ---------- | ----------
10.1.0 component template     | 59.0s   |  3.7s     |  3.1s     | 2.5s      | 3.0s      
10.1.2 component template     | 53.6s   |  3.0s     |  2.2s     | 2.2s      | 2.2s

**Comment by sharikovvladislav** [comment]
@JoostK Hi! We worked together with @nickbullock. 

I will check it soon. A bit later. We have deadline in 3 weeks, so we don't have anybody who can check this for our project. At the moment we have 10.0.5 version. So, we have to update `@angular/core` firstly. Give us some time please.

**Comment by JoostK** [comment]
@shadinamrouti Absolutely! Thanks for checking in, I'll keep the issue open.

**Comment by Splaktar** [comment]
@sharikovvladislav has your team been able to take a look at this with v10.1.2+? Based on what I've been seeing, you should see a 10-20% improvement in AOT dev mode rebuild times, but it won't yet get close to the 2-4s rebuild time that you are seeing with JIT.

I would be careful about updating all of the way to v11.1 from v10.0.5 due to https://github.com/angular/angular/issues/40635.

**Comment by JoostK** [comment]
@nickbullock Angular 11.2.5 was just released which contains #40948; that change has completely overhauled the particular code paths that you identified as being slow. I'll therefore close this issue as resolved.

@drdatarulz there could be multiple things going on there, it's hard to judge what the cause could be without more information. Ivy AOT mode enables template type checking and we have observed TypeScript bugs where certain templates take very long to type-check (which was solved in >=4.0.3). Alternatively, because you mention that `--prod` is enabled it could be that optimizers such as Angular Build Optimizer or Terser choke on something such that your build doesn't complete. I do not expect that AOT compilation itself is taking so much time, that's relatively cheap compared to the things I mentioned. Please open a new issue; ideally you are able to share a reproduction, privately if needed, as otherwise investigation is rather impossible.
